사용자 정보 CRUD api 작성

riverandeye opened the pull request:
사용자 정보에 대한 CRUD를 하는 api를 작성하였습니다. 

Post /users = 사용자 추가
Delete /users/:id = id에 해당하는 사용자 제거
Put /users/:id = id에 해당하는 사용자 정보 변경

좋은하루~

KimJye committed:
[client] | feat(search): 모바일 검색 활성화 UI 뼈대 작업

모바일 검색 활성화시 요소들에대한 화면 구성을 했습니다.
ArrawBack은 뒤로가기 버튼입니다.
InputLayer는 검색창입니다.

KimJye committed:
[client] | feat(ArrawBackSVG): 뒤로가기 SVG 추가

모바일 검색 활성화시 뒤로가기 작업의 SVG를 추가하였습니다.

KimJye committed:
[client] | feat(search): 모바일 검색 활성화 UI

디자인을 보면서, 모바일 검색 활성화의 UI를 구현했습니다.

KimJye committed:
[client] | feat(search): media로 뒤로가기버튼 display 조절

데스크톱일때는 뒤로가기버튼이 보이면 안되기때문에, 데스크톱일떄는 display none 처리를하였습니다.

KimJye committed:
[client] | feat(search): 검색 결과에 따른 주제지정

검색결과 제목 변경했습니다.

KimJye committed:
merge conflict

KimJye committed:
[client] | feat(AppBar): 모바일 검색UI 구성

모바일화면에서 검색 아이콘을 눌렀을 때, 검색 UI가 활성화됩니다.
또한, 뒤로가기 아이콘을 눌렀을 떄, 기존의 모바일 AppBar상태로
돌아갑니다.

KimJye committed:
[client] | feat(search): 모바일과 데스크탑의  검색바 UI 구현

AppBar에서 하나의 Search 컴포넌트와 상태관리를 통해 구현했습니다.

KimJye committed:
[client] | refactor(search): input width 수정

KimJye committed:
[client] | feat(search): 검색UI에 입력 후 search page 라우팅

onKeyPress의 이벤트로 검색UI에 입력 후 Enter하면useRouter를 이용하ㅕ
검색된 값을 라우팅합니다.

KimJye committed:
[client] | fix(AppBar): 모바일 검색활성화후 데스크탑 로고 없는 버그수정

모바일 검색 활성화 후, 뒤로가기 아이콘을 누르고 데스크탑으로 돌아가면
로고가 있지만, 뒤로가기 아이콘을 누르지않은 상태에서 데스크탑으로
돌아갈때 로고가 사라지고 검색UI가 맨 앞에 있는 버그가 있었습니다.

이 문제를 해결하기 위해서 media를 이용하여 데스크탑일때 Logo가
보여지도록 display block처리를 하였습니다.

KimJye committed:
[client] | refactor(search): pr 리뷰 반영

비활성화하는 함수의 이름을 동사형으로 변경하였습니다.
props를 중복해서 쓰지않기 위해destructing을 사용하였습니다.

KimJye committed:
[client] | refactor(search): 함수이름 변경

KimJye committed:
[client] | refactor(search): 기존에 작성된 routePath로 변경

Router의 pathname을 string으로 작성했었는데, 기존에 작성된 routePath의
search로 변경했습니다.

KimJye committed:
[client] | refactor(search): 함수명 변경

hyokualexkwon committed:
[client] | refactor(logo): 로고 내용 수정

hyokualexkwon committed:
[client] | refactor(VideoItem): 비디오 항목 컴포넌트 UI추가

비디오 항목 컴포넌트가 세로 카드형태 뿐만아니라 가로형 UI를 나타낼 필요가 생겼습니다. 이를 vertical, horizontal type으로 구별합니다. 모바일 화면과 데스크톱 화면 각각 UI의 type을 설정할 수 있습니다. 기본적으로 vertical type을 가지며, 필요한 경우에 horizontal type임을 명시하여 사용할 수 있습니다. 각 type의 스타일은 styled-components 에서 관리됩니다.

hyokualexkwon committed:
[client] | feat(User): 유저 페이지 화면 작성

유저 정보를 그리는 유저 페이지 화면을 작성했습니다.

KimJye committed:
[client] | refactor(SearchBar): PR리뷰 반영

함수명isActiveSearchBar에서 isSearchBarActive로 수정했습니다.
컴포넌트명을 Search에서 SearchBar로 수정했습니다.
Router대신에 useRouter로 수정했습니다.
검색 입력 부분의 마크업 구성을 변경했습니다.
input 영역이 전체 넓이를 포함하도록 수정했습니다.
불필요한 padding값들을 삭제했습니다.
SVG의 값을 동적으로 변할수있도록 수정했습니다.

KimJye committed:
[client] | feat(SearchBar): pr 리뷰 코드 수정

KimJye committed:
머지 충돌 해결

riverandeye committed:
[server] | feat(user): User dto 정의

User dto 및 UserResponseDto 를 정의하였습니다.

riverandeye committed:
[server] | feat(user): User dto 정의

User dto 및 UserResponseDto 를 정의하였습니다.

riverandeye committed:
[server] | refactor(user): Get 메소드 Param 변수명 변경

userId는 User가 아닌 다른 테이블이 가지는 Foreign key 값이기 때문에, 변수명으로 부적절하다고 생각하여 변경하였습니다.

riverandeye committed:
[server] | refactor(user): Get 메소드 Param 변수명 변경

userId는 User가 아닌 다른 테이블이 가지는 Foreign key 값이기 때문에, 변수명으로 부적절하다고 생각하여 변경하였습니다.

riverandeye committed:
[server] | feat(user): User api 추가

User api를 추가하였습니다.
Post /users = 사용자 추가
Delete /users/:id = id에 해당하는 사용자 제거
Put /users/:id = id에 해당하는 사용자 정보 변경

riverandeye committed:
[server] | feat(user): User api 추가

User api를 추가하였습니다.
Post /users = 사용자 추가
Delete /users/:id = id에 해당하는 사용자 제거
Put /users/:id = id에 해당하는 사용자 정보 변경

KimJye committed:
[client] | refactor(SearchBar): PR 리뷰 오타수정

KimJye commented on packages/server/src/user/user.controller.ts:
```diff
@@ -1,18 +1,45 @@
     return await this.userService.findVideos(userId);
   }
+
+  @Post('')
```
url 'users/'를 의미하신건가요???

riverandeye replied:
네 맞아요~

KimJye commented on packages/server/src/user/user.service.ts:
```diff
@@ -31,4 +41,23 @@ export class UserService {
+
+  public async putUser(id, putUserDto: UserDto): Promise<UserResponseDto> {
+    const user = await this.userRepository.findOne(id);
+    this.userRepository.merge(user, putUserDto);
```
배우고갑니다 👍

riverandeye replied:
감사합니다 ㅎ.ㅎ

KimJye commented on packages/server/src/user/user.controller.ts:
```diff
@@ -1,18 +1,45 @@
+
+  @Put(':id')
+  public async updateUser(
+    @Param('id') id,
```
혹시 요녀석의 type 정의를 안하신 이유가 있을까요???

riverandeye replied:
맞네요.. 리팩토링하면서 같이 지웠네요.. 수정하겠습니다.

KimJye commented on packages/server/src/user/user.service.ts:
```diff
@@ -14,12 +15,21 @@ export class UserService {
-        id,
-      },
-    });
+  private createUserResponse(user: User): UserResponseDto {
```
DB로부터 얻은 User의 정보를 객체로 가공해서 반환하는 역할이 맞나요??
혹시 맞다면 함수명에서 create부분이 헷갈릴것같습니다.

hyokualexkwon replied:
애초에 이와 같은 함수가 필요한가용? 반환할 때, `const {password, ...rest } = user;` 와 같이 필요없는 정보(password)를 제외하고 `return rest: UserResponseDto `와 같이 사용하면 안되나여?

riverandeye replied:
이부분도 많이 고민을 했는데, 생각해보니 rest 뒤에 반환형을 작성해주면 이해가 편할것 같네요.혁우님 방식 맘에드는거같아요.

Atercatus replied:
클래스에 선언된 생성자를 이용하세요.

KimJye commented on packages/server/src/user/user.service.ts:
```diff
@@ -31,4 +41,23 @@ export class UserService {
     });
   }
+
+  public async insertUser(createUserDto: UserDto): Promise<UserResponseDto> {
+    const user = this.userRepository.create(createUserDto);
+    const result = await this.userRepository.insert(user);
+    return this.createUserResponse(user);
+  }
```
이미 앞에서 id 중복을 막아주겠지만, 로직에도 user를 insert하기전에 insert하려는 user가 있는지 확인하고 insert를 하면 더 안전할 것 같아요~

KimJye commented on packages/server/src/user/user.controller.ts:
```diff
@@ -1,18 +1,45 @@
 import { Video } from '../../../typeorm/src/entity/video.entity';
+import { UserDto, UserResponseDto } from './dto';
 
 @Controller('users')
```
예외처리 부탁드립니당

KimJye committed:
Merge pull request #64 from connect-foundation/feature/search_ui_20191117

AppBar 부분 검색 UI 및 라우팅

hyokualexkwon committed:
[database] | refactor(User): 잘못된 관계 수정

user의 liked comments가 Comment를 가르키도록 entity를 수정했습니다.

hyokualexkwon committed:
[database] | refactor(Video): Video 테이블에서 like 삭제

likedUsers의 갯수를 기록하는 like 항목을 삭제합니다. 갯수를 셀때는 aggregation function을 사용하여 쿼리하도록 합니다.

hyokualexkwon committed:
[server] | refactor(): 로컬 개발 서버 포트 변경

로컬 환경에서 클라이언트는 3000에 실행하고, 서버는 4000에 실행하여 동시에 실행 시킬 수 있도록 합니다.

hyokualexkwon committed:
[server] | refactor(): 로컬 개발 서버 포트 변경

로컬 환경에서 클라이언트는 3000에 실행하고, 서버는 4000에 실행하여 동시에 실행 시킬 수 있도록 합니다.

hyokualexkwon committed:
[server] | feature(Video): 영상 리스트 API 작성

영상 리스트를 조회할 수 있는 API를 작성했습니다. 클라이언트는, page 쿼리를 통해 페이지네이션 요청을 생성할 수 있습니다. 또, sort 옵션을 통해 정렬 기준을 작성할 수 있습니다.

hyokualexkwon committed:
[server] | feature(Video): 영상 리스트 API 작성

영상 리스트를 조회할 수 있는 API를 작성했습니다. 클라이언트는, page 쿼리를 통해 페이지네이션 요청을 생성할 수 있습니다. 또, sort 옵션을 통해 정렬 기준을 작성할 수 있습니다.

hyokualexkwon committed:
[client] | refactor(VideoItem): 비디오 항목 스타일 수정

hyokualexkwon committed:
[client] | refactor(User): 비디오 chunk 로직 libs에 분리

hyokualexkwon commented on packages/server/src/user/user.service.ts:
```diff
@@ -31,4 +41,23 @@ export class UserService {
+
+  public async deleteUser(id): Promise<UserResponseDto> {
+    const user = await this.userRepository.findOne(id);
+    const result = await this.userRepository.delete(user);
```
저희 삭제 로직은 실제로 table에서 지우는게 아니라 status를 1에서 0으로 바꾸는것 아닌가요?

riverandeye replied:
좋은 지적이에요~ delete 수정하겠습니다.

Atercatus replied:
디비에 리퀘스트가 2번 날라가는데, 한번에 되지 않는게 어색합니다.

KimJye commented on packages/server/src/user/dto/user.dto.ts:
```diff
@@ -0,0 +1,7 @@
+  public readonly avatar: string;
+  public readonly email: string;
+  public readonly githubId: string;
+}
```
혹시 여기가 request에대한 Dto이신가요?
맞다면, validation을 해주시면 좋겠습니당

Atercatus replied:
`UserDto` 와 `UserResponseDto`라는 이름은 마치 추상화되어있는 `UserDto`를 `UserResponseDto`가 구현한 이름 같습니다. 네이밍을 고쳐야 할 거 같습니다.

Atercatus replied:
Controller에 직접적으로 들어오는 값에 대한 validation은 Pipe 를 이용해 주세요

KimJye commented on packages/server/src/user/user.controller.ts:
```diff
@@ -1,18 +1,45 @@
   }
+
+  @Post('')
+  public async createUser(@Body() userDto: UserDto): Promise<UserResponseDto> {
```
위 리뷰와 더불어 @Body userDto를 validation하게된다면, @UsePipe를 사용하시면 될것같습니다.
https://docs.nestjs.com/pipes 여기 참조하면 좋을것같아요~

Atercatus commented on packages/server/src/user/user.controller.ts:
```diff
@@ -1,18 +1,45 @@
-  public async getUser(@Param('userId') userId: number): Promise<User> {
-    return await this.userService.findUser(userId);
+  @Get(':id')
+  public async getUser(@Param('id') id: number): Promise<UserResponseDto> {
```
url의 값은 string 일 텐데 number로 넘겨도 이상이 없나요?

Atercatus commented on packages/server/src/user/user.controller.ts:
```diff
@@ -1,18 +1,45 @@
+    return await this.userService.findUser(id);
   }
 
   @Get(':userId/videos')
```
위에 있는 `:id` 와 `:userid`는 뭐가 다른건가요?

Atercatus commented on packages/server/src/user/user.service.ts:
```diff
@@ -14,12 +15,21 @@ export class UserService {
+      githubId: user.githubId,
+    };
+    return UserResponse;
+  }
```
```
class Exmaple {
  constructor(
    private readonly ex1: string, 
    private readonly ex2: string) {}
}

const ex: Exmaple = new Exmaple('1', '2');
```

riverandeye committed:
[docs] | docs(README): README 추가

riverandeye committed:
Merge pull request #72 from connect-foundation/feature/README_20191121

[docs] | docs(README): README 추가

hyokualexkwon committed:
[server] | refactor(Videos pipe): getVideosPipe 수정

hyokualexkwon committed:
[server] | refactor(Videos pipe): getVideosPipe 수정

hyokualexkwon committed:
[server] | refactor(video): 비디오 모듈 코드 클린 코드 적용

hyokualexkwon committed:
[server] | refactor(video): 비디오 모듈 코드 클린 코드 적용

KimJye committed:
Merge pull request #68 from connect-foundation/feature/user_detail_20191119

유저 페이지 마크업 및 스타일 작성

riverandeye committed:
[database] | feat(entity): 각 table내 count 및 popularity 컬럼 추가

One to Many, Many to Many 관계에서 매번 Join을 통해 갯수를 확인하지 않도록 컬럼에 갯수 항목들을 추가하였습니다.
또, 이러한 갯수를 통해 대상의 인기도를 평가하는 컬럼을 추가하였습니다.

riverandeye committed:
[server] | feat(entity): 각 table내 count 및 popularity 컬럼 추가

One to Many, Many to Many 관계에서 매번 Join을 통해 갯수를 확인하지 않도록 컬럼에 갯수 항목들을 추가하였습니다.
또, 이러한 갯수를 통해 대상의 인기도를 평가하는 컬럼을 추가하였습니다.

riverandeye committed:
Merge remote-tracking branch 'origin/feature/db_schema_20191122' into feature/db_schema_20191122

hyokualexkwon committed:
Merge pull request #78 from connect-foundation/feature/db_schema_20191122

[database] | feat(entity): 각 table내 count 및 popularity 컬럼 추가

hyokualexkwon committed:
Merge branch 'develop' into feature/video_list_api_20191120

hyokualexkwon committed:
Merge branch 'feature/video_list_api_20191120' of https://github.com/connect-foundation/2019-10 into feature/video_list_api_20191120

hyokualexkwon committed:
[server] | refactor(): 잘못된 모듈 경로 수정

riverandeye committed:
Merge pull request #71 from connect-foundation/feature/video_list_api_20191120

영상 리스트 API작성

riverandeye committed:
Merge branch 'feature/user_api_20191120' of https://github.com/connect-foundation/2019-10 into feature/user_api_20191120

riverandeye committed:
[server] | feat(pipes): Param에 단독으로 제공된 ID를 Validate하는 pipe 작성

User 뿐만 아니라, 각 Entity의 단일 항목을 요청할 떄 반복적으로 사용되는 pipe를 작성하였습니다.
Param으로 입력된 id에 대하여 해당 id가 Validate하는 기능을 수행합니다.

riverandeye committed:
[server] | feat(user): 사용자 정보를 Validate 하는 Pipe 작성

현재 미구현된 상태이고, Validate하는 조건들을 추후에 추가할 예정입니다.

riverandeye committed:
[server] | feat(user): user api에 필요한 dto 정의

user.dto: 사용자 생성 및 수정에 필요한 dto
user-response.dto: api 결과로 리턴되는 dto에 대한 정의
user-id-pararm.dto: url로 전달되는 param dto에 대한 정의

riverandeye committed:
[server] | feat(user): pipe를 이용한 user api 정의

각 api에 Pipe를 적용하여 입력을 validate 하였습니다.
ValidateUserPipe는 현재 완성하진 못했고, 추후 업데이트해야 합니다.
